File: experiment/ccms/src/data/block/cms_lcb.c
#include "data/block/cms_lcb.h"

/* ---- internal stream version ---- */

int cms_lcb_encode_stream(per_stream_t *s, const cms_lcb_t *v)
{
    cms_boolean_encode_stream(s, v->logEna);
    cms_object_reference_encode_stream(s, v->datSet);
    cms_trigger_conditions_encode_stream(s, v->trgOps);
    cms_int32u_encode_stream(s, v->intgPd);
    cms_object_reference_encode_stream(s, v->logRef);
    /* optFlds OPTIONAL */
    per_encode_optional(s, v->optFlds_present);
    if (v->optFlds_present)
        cms_lcb_opt_flds_encode_stream(s, v->optFlds);
    /* bufTm OPTIONAL */
    per_encode_optional(s, v->bufTm_present);
    if (v->bufTm_present)
        cms_int32u_encode_stream(s, v->bufTm);
    return CMS_OK;
}

int cms_lcb_decode_stream(per_stream_t *s, cms_lcb_t *v)
{
    cms_boolean_decode_stream(s, &v->logEna);
    cms_object_reference_decode_stream(s, v->datSet);
    cms_trigger_conditions_decode_stream(s, v->trgOps);
    cms_int32u_decode_stream(s, &v->intgPd);
    cms_object_reference_decode_stream(s, v->logRef);
    /* optFlds OPTIONAL */
    v->optFlds_present = per_decode_optional(s);
    if (v->optFlds_present)
        cms_lcb_opt_flds_decode_stream(s, v->optFlds);
    /* bufTm OPTIONAL */
    v->bufTm_present = per_decode_optional(s);
    if (v->bufTm_present)
        cms_int32u_decode_stream(s, &v->bufTm);
    return CMS_OK;
}

/* ---- public buffer version ---- */

CMS_EXPORT int cms_lcb_encode(const cms_lcb_t *v, uint8_t *b, int *l)
{
    per_stream_t w;
    per_stream_init_write(&w, b, (size_t)*l);
    cms_lcb_encode_stream(&w, v);
    *l = (int)per_stream_bytes_written(&w);
    return CMS_OK;
}

CMS_EXPORT int cms_lcb_decode(const uint8_t *b, int l, cms_lcb_t *v)
{
    per_stream_t r;
    per_stream_init_read(&r, b, (size_t)l);
    cms_lcb_decode_stream(&r, v);
    return CMS_OK;
}

File: old/jcms/src/main/java/com/ysh/jcms/datatypes2/data/block/CmsLcb.java
package com.ysh.jcms.datatypes2.data.block;

import com.sun.jna.Structure;
import com.ysh.jcms.datatypes2.data.basic.CmsBoolean;
import com.ysh.jcms.datatypes2.data.basic.CmsInt32U;

import java.util.Arrays;
import java.util.List;

/**
 * cms_lcb_t — LCB (Log Control Block).
 *
 * C: typedef struct {
 *     int         logEna;
 *     char        datSet[256];
 *     uint8_t     trgOps[1];
 *     uint32_t    intgPd;
 *     char        logRef[256];
 *     uint8_t     optFlds[1];
 *     int         optFlds_present;
 *     uint32_t    bufTm;
 *     int         bufTm_present;
 * } cms_lcb_t;
 */
public class CmsLcb extends Structure {
    public CmsBoolean logEna = new CmsBoolean();
    public byte[] datSet = new byte[256];
    public byte[] trgOps = new byte[1];
    public CmsInt32U intgPd = new CmsInt32U();
    public byte[] logRef = new byte[256];
    public byte[] optFlds = new byte[1];
    public CmsBoolean optFldsPresent = new CmsBoolean();
    public CmsInt32U bufTm = new CmsInt32U();
    public CmsBoolean bufTmPresent = new CmsBoolean();

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("logEna", "datSet", "trgOps", "intgPd", "logRef",
                "optFlds", "optFldsPresent", "bufTm", "bufTmPresent");
    }
}
